FirebaseSessions/Sources/Identifiers.swift

```diff
@@ -0,0 +1,106 @@
+//
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import FirebaseInstallations
+
+let sessionIDUserDefaultsKey = "com.firebase.sessions.sessionID"
+let lastSessionIDUserDefaultsKey = "com.firebase.sessions.lastSessionID"
+
+protocol IdentifierProvider {
+  var installationID: String {
+    get
+  }
+
+  var sessionID: String {
+    get
+  }
+
+  var lastSessionID: String {
+    get
+  }
+}
+
+///
+/// Identifiers is responsible for:
+///   1) Getting the Installation ID from Installations
+///   2) Generating the Session ID
+///   3) Persisting and reading the Session ID from the last session
+///   (Maybe) 4) Persisting, reading, and incrementing an increasing index
+///
+class Identifiers: IdentifierProvider {
+  private let installationIDProvider: InstallationIDProvider
+
+  private var uuid: UUID
+
+  init(installationIDProvider: InstallationIDProvider) {
+    self.installationIDProvider = installationIDProvider
+    uuid = UUID()
+  }
+
+  func generateNewSessionID() {
+    uuid = UUID()
+
+    let lastStoredSessionID = UserDefaults.standard.string(forKey: sessionIDUserDefaultsKey) ?? ""
+    UserDefaults.standard.set(lastStoredSessionID, forKey: lastSessionIDUserDefaultsKey)
+
+    let newSessionID = uuid.uuidString.replacingOccurrences(of: "-", with: "").lowercased()
+    UserDefaults.standard.set(newSessionID, forKey: sessionIDUserDefaultsKey)
+  }
+
+  // This method must be run on a background thread due to how Firebase Installations
+  // handles threading.
+  var installationID: String {
```

Comment:
I'm not sure we should try to make this getter behave synchronously. I think we should instead replace this property with a method that takes a completion handler to get the block.

Reply:
That would make using the api much harder though no? At some point I need an ugly semaphor and I'd rather not offload that to other classes

Reply:
I would just keep the API surface asynchronous with completion handlers and avoid using semaphores throughout the SDK.

Currently, accessing this property will call to installations every time. How many times in the SDK's lifecycle will we need an installation ID? Only once at every startup of the SDK?